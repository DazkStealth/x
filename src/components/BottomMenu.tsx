import React from 'react';
import styled from 'styled-components';
import { useNavigate } from 'react-router-dom';
import { MdAccountBalance, MdApps } from 'react-icons/md';
import { PiPaperPlaneTiltFill } from 'react-icons/pi';
import { FaHistory } from "react-icons/fa";
import { usePrivy } from '@privy-io/react-auth';

import { navigationRoute } from '../navigation';

const BottomMenu = () => {
  const { authenticated } = usePrivy();
  const navigate = useNavigate();

  if (!authenticated) return null;

  return (
    <Wrapper>
      <MenuItem onClick={() => navigate(navigationRoute.home)}>
        <MdAccountBalance size={30} />
      </MenuItem>
      <MenuItem>
        <PiPaperPlaneTiltFill size={26} />
      </MenuItem>
      <MenuItem>
        <FaHistory size={26} />
      </MenuItem>
      <MenuItem>
        <MdApps size={30} />
      </MenuItem>
    </Wrapper>
  );
}

const Wrapper = styled.div`
  max-width: 800px;
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 40px;
  position: fixed;
  bottom: 60px;
  left: 50%;
  transform: translateX(-50%);
`;

const MenuItem = styled.span`
  display: flex;
  align-items: center;
  justify-content: center;
  background: ${({ theme }) => theme.color.background.bottomMenuItem};
  color: ${({ theme }) => theme.color.text.bottomMenuItem};
  border-radius: 16px;
  height: 60px;
  width: 60px;
  overflow: hidden;
  box-shadow: 7px 9px 59px 4px rgba(0,0,0,0.45);
  cursor: pointer;
  
  &:hover {
    opacity: 0.5;
  }
  
  &:active {
    opacity: 0.2;
  }
`;

export default BottomMenu;
