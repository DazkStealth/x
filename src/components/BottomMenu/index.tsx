import React from 'react';
import styled from 'styled-components';
import { useLocation, useNavigate } from 'react-router-dom';
import { usePrivy } from '@privy-io/react-auth';
import { useTranslation } from 'react-i18next';

// images
import { ReactComponent as IconSend } from '../../assets/images/icon/icon-send.svg';
import { ReactComponent as IconHistory } from '../../assets/images/icon/icon-history.svg';
import { ReactComponent as IconApps } from '../../assets/images/icon/icon-apps.svg';
import { ReactComponent as IconWallet } from '../../assets/images/icon/icon-wallet.svg';
import { ReactComponent as IconHome } from '../../assets/images/icon/icon-home.svg';

// navigation
import { navigationRoute } from '../../navigation';

// hooks
import useBottomMenuModal from '../../hooks/useBottomMenuModal';

const BottomMenu = () => {
  const { authenticated } = usePrivy();
  const navLocation = useLocation();
  const navigate = useNavigate();
  const [t] = useTranslation();
  const { setActiveMenuItemIndex, activeMenuItemIndex } = useBottomMenuModal();

  if (!authenticated) return null;

  const mainMenuItems = [
    {
      icon: <IconSend />,
      label: t`menuAction.send`,
    },
    {
      icon: <IconHistory />,
      label: t`menuAction.history`,
    },
    {
      icon: <IconWallet />,
      label: t`menuAction.account`,
    },
    {
      icon: <IconApps />,
      label: t`menuAction.apps`,
    },
  ];

  const isHomeActive = activeMenuItemIndex === null
    && navLocation.pathname === '/';

  return (
    <Wrapper>
      <HomeMenuItem>
        <MenuItem
          onClick={() => navigate(navigationRoute.home)}
          className={isHomeActive ? 'active' : ''}
        >
          <IconHome />
        </MenuItem>
      </HomeMenuItem>
      <MainMenuItems>
        {mainMenuItems.map((item, index) => (
          <MenuItem
            key={item.label + index}
            onClick={() => {
              // TODO: replace this when apps modal design is ready
              if (index === 3) {
                navigate(navigationRoute.apps);
                return;
              }

              if (activeMenuItemIndex === index) {
                // toggle out if already active
                setActiveMenuItemIndex(null);
                return;
              }

              setActiveMenuItemIndex(index);
            }}
            className={activeMenuItemIndex === index ? 'active' : ''}
          >
            {item.icon}
            <span>{item.label}</span>
          </MenuItem>
        ))}
      </MainMenuItems>
    </Wrapper>
  );
}

const MenuItem = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 12px;
  color: ${({ theme }) => theme.color.text.bottomMenuItem};
  height: 48px;
  overflow: hidden;
  cursor: pointer;
  transition: all .1s ease-in-out;
  font-weight: 400;
  letter-spacing: -0.5px;
  font-size: 14px;

  &:first-child {
    margin-left: 17px;
  }

  &:last-child {
    margin-right: 17px;
  }

  span {
    display: none;
  }
`;

const Wrapper = styled.div`
  max-width: 800px;
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 10px;
  position: fixed;
  bottom: 40px;
  left: 50%;
  z-index: 100;
  transform: translateX(-50%);

  &:not(&:hover) ${MenuItem}.active, ${MenuItem}:hover {
    padding: 0 13px;
    margin: 0;
    border-radius: 100px;
    background: ${({ theme }) => theme.color.background.bottomMenuItemHover};

    span {
      display: block;
    }
  }
`;

const MainMenuItems = styled.div`
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: space-between;
  backdrop-filter: blur(5px);
  background: ${({ theme }) => theme.color.background.bottomMenu};
  border-radius: 130px;
  padding: 6px 7px;
  width: 280px;
  height: 60px;
`;

const HomeMenuItem = styled(MainMenuItems)`
  flex: none;
  width: 60px;
  height: 60px;

  ${MenuItem} {
    width: 48px;

    &:first-child, &:last-child {
      margin: 0
    }

    &:hover {
      padding: 0;
    }
  }
`;

export default BottomMenu;
