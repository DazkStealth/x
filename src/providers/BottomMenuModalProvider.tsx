import React, { createContext, useMemo } from 'react';
import { usePrivy } from '@privy-io/react-auth';

// components
import BottomMenuModal from '../components/BottomMenuModal';

export interface BottomMenuModalContext {
  data: {
    activeMenuItemIndex: number | null;
    setActiveMenuItemIndex: (index: number | null) => void;
  }
}

export const ProviderContext = createContext<BottomMenuModalContext | null>(null);

const BottomMenuModalProvider = ({ children }: React.PropsWithChildren) => {
  const { authenticated } = usePrivy();
  const [activeMenuItemIndex, setActiveMenuItemIndex] = React.useState<number | null>(null);

  const contextData = useMemo(() => ({
    activeMenuItemIndex,
    setActiveMenuItemIndex,
  }), [activeMenuItemIndex]);

  return (
    <ProviderContext.Provider value={{ data: contextData }}>
      {children}
      {authenticated && (
        <BottomMenuModal
          activeMenuItemIndex={activeMenuItemIndex}
          onClose={() => setActiveMenuItemIndex(null)}
        />
      )}
    </ProviderContext.Provider>
  );
}

export default BottomMenuModalProvider;
