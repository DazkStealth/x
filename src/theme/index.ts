export interface Theme {
  font: {
    primary: string;
  }
  color: {
    background: {
      body: string;
      bottomMenu: string;
      bottomMenuItem: string;
      loadingLogo: string;
      buttonPrimary: string;
    },
    text: {
      body: string;
      bottomMenuItem: string;
      bottomMenuItemActive: string;
      loadingLogo: string;
      buttonPrimary: string;
    },
    border: {
      bottomMenuItemBottomActive: string;
    }
  }
}

export const defaultTheme: Theme = {
  font: {
    primary: '"Roboto", sans',
  },
  color: {
    background: {
      body: '#24142f',
      bottomMenu: 'rgba(50,50,66,0.7)',
      bottomMenuItem: '#444d55',
      loadingLogo: '#fff',
      buttonPrimary: '#c63bea',
    },
    text: {
      body: '#fff',
      bottomMenuItem: '#fff',
      bottomMenuItemActive: '#444d55',
      loadingLogo: '#997cfa',
      buttonPrimary: '#fff',
    },
    border: {
      bottomMenuItemBottomActive: '#fff',
    }
  },
};
