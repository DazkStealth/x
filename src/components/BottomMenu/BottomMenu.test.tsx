import renderer, { ReactTestRendererJSON } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter } from 'react-router-dom';
import { usePrivy } from '@privy-io/react-auth';

// components
import BottomMenu from './';

// theme
import { defaultTheme } from '../../theme';

// providers
import BottomMenuModalProvider from '../../providers/BottomMenuModalProvider';
import LanguageProvider from '../../providers/LanguageProvider';

describe('<BottomMenu />', () => {
  it('renders correctly', () => {
    (usePrivy as jest.Mock).mockImplementation(() => ({ authenticated: true }));
    const tree = renderer
      .create(
        <BrowserRouter>
          <ThemeProvider theme={defaultTheme}>
            <LanguageProvider>
              <BottomMenuModalProvider>
                <BottomMenu />
              </BottomMenuModalProvider>
            </LanguageProvider>
          </ThemeProvider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

    const treeElements = tree as ReactTestRendererJSON[];
    const [bottomMenuElement] = treeElements;
    expect((bottomMenuElement.children?.[0] as ReactTestRendererJSON)?.children?.length).toBe(1); // menu home button
    expect((bottomMenuElement.children?.[1] as ReactTestRendererJSON)?.children?.length).toBe(4); // other menu items
    expect(bottomMenuElement.type).toBe('div');
    expect(bottomMenuElement?.children?.[0]).toHaveStyleRule('background', defaultTheme.color.background.bottomMenu);
    expect(bottomMenuElement?.children?.[1]).toHaveStyleRule('background', defaultTheme.color.background.bottomMenu);
    jest.clearAllMocks();
  });

  it('renders correctly if not authenticated', () => {
    (usePrivy as jest.Mock).mockImplementation(() => ({ authenticated: false }));
    const tree = renderer
      .create(
        <BrowserRouter>
          <ThemeProvider theme={defaultTheme}>
            <LanguageProvider>
              <BottomMenuModalProvider>
                <BottomMenu />
              </BottomMenuModalProvider>
            </LanguageProvider>
          </ThemeProvider>
        </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

    const treeElement = tree as ReactTestRendererJSON;
    expect(treeElement).toBe(null);
    jest.clearAllMocks();
  });
});


