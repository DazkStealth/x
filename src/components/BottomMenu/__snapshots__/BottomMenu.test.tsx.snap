// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BottomMenu /> renders correctly 1`] = `
[
  .c4 {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 12px;
  color: #fff;
  height: 48px;
  overflow: hidden;
  cursor: pointer;
  transition: all .1s ease-in-out;
  font-weight: 400;
  letter-spacing: -0.5px;
  font-size: 14px;
}

.c4:first-child {
  margin-left: 17px;
}

.c4:last-child {
  margin-right: 17px;
}

.c4 span {
  display: none;
}

.c0 {
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: space-between;
  gap: 10px;
  position: fixed;
  bottom: 40px;
  left: 50%;
  z-index: 100;
  transform: translateX(-50%);
}

.c0:not(.c0:hover) .c3.active,
.c0 .c3:hover {
  padding: 0 13px;
  margin: 0;
  border-radius: 100px;
  background: rgba(216, 232, 255, 0.10);
}

.c0:not(.c0:hover) .c3.active span,
.c0 .c3:hover span {
  display: block;
}

.c1 {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: space-between;
  backdrop-filter: blur(5px);
  background: rgba(16,16,16,0.7);
  border-radius: 130px;
  padding: 6px 7px;
  width: 280px;
  height: 60px;
}

.c2 {
  flex: none;
  width: 60px;
  height: 60px;
}

.c2 .c3 {
  width: 48px;
}

.c2 .c3:first-child,
.c2 .c3:last-child {
  margin: 0;
}

.c2 .c3:hover {
  padding: 0;
}

<div
    className="c0"
  >
    <div
      className="c1 c2"
    >
      <div
        className="c3 c4 active"
        onClick={[Function]}
      >
        <svg />
      </div>
    </div>
    <div
      className="c1"
    >
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          Send
        </span>
      </div>
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          History
        </span>
      </div>
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          Account
        </span>
      </div>
      <div
        className="c3 c4"
        onClick={[Function]}
      >
        <svg />
        <span>
          Apps
        </span>
      </div>
    </div>
  </div>,
  .c20 {
  font-size: 15px;
  font-weight: 700;
  padding: 15px 45px;
  border: none;
  border-radius: 100px;
  background: #63626B;
  color: #1D1D1D;
  transition: all 0.2s ease-in-out;
  margin-bottom: 15px;
  width: 100%;
}

.c20:focus {
  outline: none;
}

.c17 {
  font-size: 14px;
  font-weight: 400;
}

.c25 {
  display: none;
  max-width: 100%;
  border-radius: 20px;
}

.c24 {
  animation: hLhBjP 1s linear infinite alternate;
  height: 100%;
  border-radius: 20px;
}

.c8 {
  padding: 17px;
  border-radius: 10px;
  background: rgba(217, 217, 217, 0.05);
  color: #fff;
  width: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  user-select: none;
}

.c9 {
  background: transparent;
  color: #fff;
  border: none;
  font-size: 14px;
  font-weight: 400;
  flex: 1;
}

.c9:focus {
  outline: none;
}

.c9::placeholder {
  color: #838383;
}

.c7 {
  font-size: 15px;
  font-weight: 700;
  margin-bottom: 24px;
}

.c6 {
  display: flex;
  flex-direction: column;
  margin-bottom: 24px;
}

.c10 {
  opacity: 0.2;
  background: linear-gradient(270deg, rgba(217, 217, 217, 0.00) 0%, #D9D9D9 50.52%, rgba(217, 217, 217, 0.00) 100%);
  height: 1px;
  width: 100%;
}

.c13 {
  width: 42px;
  height: 42px;
  border-radius: 50%;
}

.c12 {
  width: 42px;
}

.c14 {
  flex: 1;
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c15 {
  font-size: 15px;
  font-weight: 700;
  word-break: break-word;
}

.c16 {
  font-size: 14px;
  font-weight: 400;
  word-break: break-word;
}

.c11 {
  border-radius: 10px;
  background: rgba(217, 217, 217, 0.05);
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-content: center;
  gap: 12px;
  padding: 11px 13px;
  cursor: pointer;
  user-select: none;
  margin-bottom: 10px;
}

.c11:hover {
  background: rgba(217, 217, 217, 0.2);
}

.c5 {
  width: 100%;
  max-height: calc(100vh - 240px);
  overflow: hidden;
  min-height: 100%;
  padding-bottom: 150px;
  overflow-y: scroll;
  -ms-overflow-style: none;
  scrollbar-width: none;
}

.c5::-webkit-scrollbar {
  display: none;
}

.c19 {
  position: absolute;
  bottom: 0;
  padding: 34px 17px 22px;
  left: 0;
  width: 100%;
  background: linear-gradient(180deg, rgba(16, 16, 16, 0.00) 0%, rgba(16, 16, 16, 0.90) 52.08%);
}

.c18 {
  padding: 0 17px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-content: center;
  margin-top: 14px;
  color: rgba(255, 255, 255, 0.3);
  user-select: none;
  width: 100%;
}

.c21 {
  width: 100%;
  word-break: break-all;
  text-align: center;
}

.c22 {
  display: flex;
  gap: 15px;
  flex-direction: row;
  align-content: flex-start;
  justify-content: flex-start;
  flex-wrap: wrap;
  width: 100%;
  min-height: 100%;
}

.c23 {
  width: 90px;
  height: 90px;
  border-radius: 20px;
  overflow: hidden;
  transition: all .1s ease-in-out;
  cursor: pointer;
}

.c23:hover {
  transform: scale(1.1);
  opacity: 0.9;
}

.c0 {
  position: fixed;
  z-index: 98;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  transition: 100ms linear;
  backdrop-filter: blur(0px);
  display: none;
  align-items: flex-end;
  justify-content: center;
  padding-bottom: 114px;
  overflow: hidden;
}

.c1 {
  transition: 100ms linear;
  transform: translateY(1000px);
  display: flex;
  flex-direction: row;
  align-content: start;
  justify-content: start;
  width: 100%;
}

.c2 {
  align-self: flex-start;
  transform: translateX(calc(0 * -100vw));
  display: flex;
  flex-direction: row;
  align-content: start;
  justify-content: start;
}

.c3 {
  width: 100vw;
  display: flex;
  align-content: center;
  justify-content: center;
}

.c4 {
  width: 350px;
  min-height: 350px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border-radius: 32px;
  padding: 34px 23px;
  background: rgba(16, 16, 16, 0.70);
  backdrop-filter: blur(5px);
  overflow: hidden;
}

<div
    className="c0"
  >
    <div
      className="c1"
    >
      <div
        className="c2"
      >
        <div
          className="c3"
        >
          <div
            className="c4"
          >
            <div
              className="c5"
            >
              <div
                className="c6"
              >
                <label
                  className="c7"
                >
                  Choose recipient
                </label>
                <div
                  className="c8"
                >
                  <input
                    className="c9"
                    onChange={[Function]}
                    placeholder="To: name, address"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className="c6"
              >
                <div
                  className="c10"
                />
              </div>
              <div
                className="c6"
              >
                <label
                  className="c7"
                >
                  Choose token
                </label>
                <div>
                  <div
                    className="c11"
                    onClick={[Function]}
                  >
                    <div
                      className="c12"
                    >
                      <img
                        alt="Ether"
                        className="c13"
                        src="https://public.etherspot.io/buidler/chain_logos/ethereum.png"
                        title="Ether"
                      />
                    </div>
                    <div
                      className="c14"
                    >
                      <p
                        className="c15"
                      >
                        Ether
                      </p>
                      <p
                        className="c16"
                      >
                        0 ETH
                      </p>
                    </div>
                  </div>
                  <div
                    className="c11"
                    onClick={[Function]}
                  >
                    <div
                      className="c14"
                    >
                      <p
                        className="c15"
                      >
                        tk1
                      </p>
                      <p
                        className="c16"
                      >
                        0 TK1
                      </p>
                    </div>
                  </div>
                  <div
                    className="c11"
                    onClick={[Function]}
                  >
                    <div
                      className="c14"
                    >
                      <p
                        className="c15"
                      >
                        tk2
                      </p>
                      <p
                        className="c16"
                      >
                        0 TK2
                      </p>
                    </div>
                  </div>
                  <div
                    className="c11"
                    onClick={[Function]}
                  >
                    <div
                      className="c14"
                    >
                      <p
                        className="c15"
                      >
                        tk3
                      </p>
                      <p
                        className="c16"
                      >
                        0 TK3
                      </p>
                    </div>
                  </div>
                </div>
              </div>
              <div
                className="c6"
              >
                <label
                  className="c7"
                >
                  Enter amount
                </label>
                <div
                  className="c8"
                >
                  <input
                    className="c9"
                    disabled={true}
                    onChange={[Function]}
                    placeholder="0.00"
                    type="text"
                    value=""
                  />
                  <p
                    className="c17"
                  />
                </div>
                <div
                  className="c18"
                >
                  <p
                    className="c17"
                  >
                    0.00 left
                  </p>
                </div>
              </div>
              <div
                className="c19"
              >
                <button
                  className="c20"
                  disabled={true}
                  onClick={[Function]}
                >
                  Send
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          className="c3"
        >
          <div
            className="c4"
          >
            <div>
              History
            </div>
          </div>
        </div>
        <div
          className="c3"
        >
          <div
            className="c4"
          >
            <div
              className="c21"
            >
              <p
                className=""
              >
                0x7F30B1960D5556929B03a0339814fE903c55a347
              </p>
            </div>
          </div>
        </div>
        <div
          className="c3"
        >
          <div
            className="c4"
          >
            <div
              className="c22"
            >
              <div
                className="c23"
                onClick={[Function]}
              >
                <div
                  className="c24"
                />
                <img
                  className="c25"
                  src="icon.png"
                />
              </div>
              <div
                className="c23"
                onClick={[Function]}
              >
                <div
                  className="c24"
                />
                <img
                  className="c25"
                  src="icon.png"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`<BottomMenu /> renders correctly if not authenticated 1`] = `null`;
